File: Services/OpenAIService.cs
using System.Text;
using OpenAI.Chat;
using WriteCommit.Constants;
using WriteCommit.Models;

namespace WriteCommit.Services;

public class OpenAIService
{
    private readonly string _apiKey;
    private readonly string _patternsDirectory;
    private const int MaxContextTokens = 128000;

    public OpenAIService(string apiKey)
    {
        if (string.IsNullOrEmpty(apiKey))
        {
            throw new ArgumentNullException(nameof(apiKey), "API key cannot be null or empty");
        }

        _apiKey = apiKey;
        _patternsDirectory = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "patterns");
    }

    public async Task<string> GenerateCommitMessageAsync(
        List<DiffChunk> chunks,
        string pattern,
        int temperature,
        int topP,
        int presence,
        int frequency,
        string model,
        bool verbose
    )
    {
        if (chunks.Count == 1)
        {
            // Single chunk - process normally
            return await ProcessSingleChunkAsync(
                chunks[0],
                pattern,
                temperature,
                topP,
                presence,
                frequency,
                model,
                verbose
            );
        }
        else
        {
            // Multiple chunks - process individually and combine
            var commitMessages = new List<string>();
            if (verbose)
            {
                Console.WriteLine($"Processing {chunks.Count} chunks in parallel...");
            }

            // Create tasks for parallel processing
            var chunkTasks = chunks
                .Select(
                    async (chunk, index) =>
                    {
                        if (verbose)
                        {
                            Console.WriteLine(
                                $"Processing chunk {index + 1}/{chunks.Count}: {chunk.FileName}"
                            );
                        }

                        var chunkMessage = await ProcessSingleChunkAsync(
                            chunk,
                            PatternNames.ChunkPattern,
                            temperature,
                            topP,
                            presence,
                            frequency,
                            model,
                            verbose
                        );

                        return new
                        {
                            Index = index,
                            FileName = chunk.FileName,
                            Message = chunkMessage,
                        };
                    }
                )
                .ToArray();

            // Wait for all tasks to complete
            var results = await Task.WhenAll(chunkTasks);

            // Process results in order
            foreach (var result in results.OrderBy(r => r.Index))
            {
                if (!string.IsNullOrWhiteSpace(result.Message))
                {
                    commitMessages.Add(
                        $"Chunk {result.Index + 1} ({result.FileName}): {result.Message.Trim()}"
                    );
                }
            }

            // Combine all chunk messages into a coherent commit message
            if (commitMessages.Count == 0)
            {
                throw new InvalidOperationException(
                    "Failed to generate commit message from any chunk"
                );
            }

            // For multiple chunks, create a summary message
            var combinedMessage = await CombineChunkMessagesAsync(
                commitMessages,
                pattern,
                temperature,
                topP,
                presence,
                frequency,
                model,
                verbose
            );
            return combinedMessage;
        }
    }

    private async Task<string> ProcessSingleChunkAsync(
        DiffChunk chunk,
        string pattern,
        int temperature,
        int topP,
        int presence,
        int frequency,
        string model,
        bool verbose
    )
    {
        if (verbose)
        {
            Console.WriteLine($"Processing chunk with OpenAI API using pattern: {pattern}");
        }

        // Load the system prompt from the pattern file
        var systemPrompt = await LoadPatternAsync(pattern);
        if (string.IsNullOrWhiteSpace(systemPrompt))
        {
            throw new InvalidOperationException($"Failed to load pattern: {pattern}");
        }

        // Create a client for this specific model
        var chatClient = new ChatClient(model, _apiKey);

        // Create the chat messages
        var messages = new List<ChatMessage>
        {
            new SystemChatMessage(systemPrompt),
            new UserChatMessage(chunk.Content),
        };

        // Create chat completion options
        var options = new ChatCompletionOptions
        {
            Temperature = ConvertTemperature(temperature),
            TopP = ConvertTopP(topP),
            PresencePenalty = ConvertPenalty(presence),
            FrequencyPenalty = ConvertPenalty(frequency),
        };

        if (verbose)
        {
            Console.WriteLine($"Sending request to OpenAI API with model: {model}");
        }

        try
        {
            // Send request to OpenAI
            var response = await chatClient.CompleteChatAsync(messages, options);

            if (verbose)
            {
                Console.WriteLine("Received response from OpenAI API");
            }

            if (response.Value != null)
            {
                return response.Value.Content[0].Text.Trim();
            }
            else
            {
                throw new InvalidOperationException("No response received from OpenAI API");
            }
        }
        catch (Exception ex)
        {
            if (verbose)
            {
                Console.WriteLine($"Error calling OpenAI API: {ex.Message}");
            }
            throw new InvalidOperationException(
                $"Failed to generate commit message: {ex.Message}",
                ex
            );
        }
    }

    private async Task<string> CombineChunkMessagesAsync(
        List<string> chunkMessages,
        string pattern,
        int temperature,
        int topP,
        int presence,
        int frequency,
        string model,
        bool verbose
    )
    {
        if (verbose)
        {
            Console.WriteLine("Combining chunk messages into final commit message...");
        }

        // Load the system prompt from the pattern file
        var systemPrompt = await LoadPatternAsync(pattern);
        if (string.IsNullOrWhiteSpace(systemPrompt))
        {
            throw new InvalidOperationException($"Failed to load pattern: {pattern}");
        }

        var combinedContent = string.Join("\n\n", chunkMessages);
        var estimatedTokens = EstimateTokenCount(systemPrompt) + EstimateTokenCount(combinedContent);

        if (estimatedTokens > MaxContextTokens && chunkMessages.Count > 1)
        {
            if (verbose)
            {
                Console.WriteLine("Context length exceeded, re-chunking summaries...");
            }

            var groupedSummaries = new List<string>();
            var currentGroup = new List<string>();
            var currentTokens = EstimateTokenCount(systemPrompt);

            foreach (var msg in chunkMessages)
            {
                var msgTokens = EstimateTokenCount(msg);
                if (currentTokens + msgTokens > MaxContextTokens / 2 && currentGroup.Count > 0)
                {
                    var summary = await CombineChunkMessagesAsync(currentGroup, pattern, temperature, topP, presence, frequency, model, verbose);
                    groupedSummaries.Add(summary);
                    currentGroup.Clear();
                    currentTokens = EstimateTokenCount(systemPrompt);
                }

                currentGroup.Add(msg);
                currentTokens += msgTokens;
            }

            if (currentGroup.Count > 0)
            {
                var summary = await CombineChunkMessagesAsync(currentGroup, pattern, temperature, topP, presence, frequency, model, verbose);
                groupedSummaries.Add(summary);
            }

            return await CombineChunkMessagesAsync(groupedSummaries, pattern, temperature, topP, presence, frequency, model, verbose);
        }

        // Create a client for this specific model
        var chatClient = new ChatClient(model, _apiKey);

        // Create the chat messages
        var messages = new List<ChatMessage>
        {
            new SystemChatMessage(systemPrompt),
            new UserChatMessage(combinedContent),
        };

        // Create chat completion options
        var options = new ChatCompletionOptions
        {
            Temperature = ConvertTemperature(temperature),
            TopP = ConvertTopP(topP),
            PresencePenalty = ConvertPenalty(presence),
            FrequencyPenalty = ConvertPenalty(frequency),
        };

        try
        {
            // Send request to OpenAI
            var response = await chatClient.CompleteChatAsync(messages, options);

            if (response.Value != null)
            {
                return response.Value.Content[0].Text.Trim();
            }
            else
            {
                throw new InvalidOperationException("No response received from OpenAI API");
            }
        }
        catch (Exception ex)
        {
            // If combining fails, return the first chunk message as fallback
            if (verbose)
            {
                Console.WriteLine($"Error calling OpenAI API: {ex.Message}");
                Console.WriteLine("Failed to combine messages, using first chunk as fallback");
            }
            return chunkMessages[0].Split(':').Skip(1).FirstOrDefault()?.Trim() ?? chunkMessages[0];
        }
    }

    /// <summary>
    /// Loads a pattern file from the patterns directory
    /// </summary>
    private async Task<string> LoadPatternAsync(string patternName)
    {
        var patternPath = Path.Combine(_patternsDirectory, patternName, "system.md");

        if (!File.Exists(patternPath))
        {
            throw new FileNotFoundException($"Pattern file not found: {patternPath}");
        }

        return await File.ReadAllTextAsync(patternPath);
    }

    /// <summary>
    /// Converts the temperature value from the CLI (0-2) to OpenAI API range (0-2)
    /// </summary>
    private float ConvertTemperature(int temperature)
    {
        // OpenAI uses 0-2 for temperature
        return Math.Clamp(Convert.ToSingle(temperature), 0f, 2f);
    }

    /// <summary>
    /// Converts the top-p value from the CLI (0-1) to OpenAI API range (0-1)
    /// </summary>
    private float ConvertTopP(int topP)
    {
        // OpenAI uses 0-1 for top-p
        return Math.Clamp((float)topP, 0f, 1f);
    }

    /// <summary>
    /// Converts the penalty value from the CLI (-2 to 2) to OpenAI API range (-2 to 2)
    /// </summary>
    private float ConvertPenalty(int penalty)
    {
        // OpenAI uses -2 to 2 for penalties
        return Math.Clamp((float)penalty, -2f, 2f);
    }

    /// <summary>
    /// Estimates token count using a rough 4 chars per token heuristic
    /// </summary>
    private int EstimateTokenCount(string text)
    {
        return Math.Max(1, text.Length / 4);
    }
}

File: src/services/OpenAIService.ts
import OpenAI from 'openai';
import { promises as fs } from 'fs';
import { join } from 'path';
import { DiffChunk } from '../models';
import { PatternNames } from '../constants';

export class OpenAIService {
  private readonly apiKey: string;
  private readonly endpoint: string;
  private readonly useAzure: boolean;
  private readonly patternsDirectory: string;
  private static readonly MAX_CONTEXT_TOKENS = 128000;
  private readonly openai: OpenAI;

  constructor(apiKey: string, endpoint?: string, useAzure = false) {
    if (!apiKey) {
      throw new Error('API key cannot be null or empty');
    }

    this.apiKey = apiKey;
    this.endpoint = endpoint || 'https://api.openai.com/v1';
    this.useAzure = useAzure;
    this.patternsDirectory = join(__dirname, '..', '..', 'patterns');

    // Create OpenAI client
    this.openai = new OpenAI({
      apiKey: this.apiKey,
      baseURL: this.endpoint.endsWith('/v1') ? this.endpoint : `${this.endpoint}/v1`,
    });
  }

  async generateCommitMessage(
    chunks: DiffChunk[],
    pattern: string,
    temperature: number,
    topP: number,
    presence: number,
    frequency: number,
    model: string,
    verbose: boolean
  ): Promise<string> {
    if (chunks.length === 1) {
      // Single chunk - process normally
      return await this.processSingleChunk(
        chunks[0],
        pattern,
        temperature,
        topP,
        presence,
        frequency,
        model,
        verbose
      );
    } else {
      // Multiple chunks - process individually and combine
      const commitMessages: string[] = [];
      if (verbose) {
        console.log(`Processing ${chunks.length} chunks in parallel...`);
      }

      // Create tasks for parallel processing
      const chunkTasks = chunks.map(async (chunk, index) => {
        if (verbose) {
          console.log(`Processing chunk ${index + 1}/${chunks.length}: ${chunk.fileName}`);
        }

        const chunkMessage = await this.processSingleChunk(
          chunk,
          PatternNames.CHUNK_PATTERN,
          temperature,
          topP,
          presence,
          frequency,
          model,
          verbose
        );

        return {
          index,
          fileName: chunk.fileName,
          message: chunkMessage,
        };
      });

      // Wait for all tasks to complete
      const results = await Promise.all(chunkTasks);

      // Process results in order
      for (const result of results.sort((a, b) => a.index - b.index)) {
        if (result.message && result.message.trim()) {
          commitMessages.push(`Chunk ${result.index + 1} (${result.fileName}): ${result.message.trim()}`);
        }
      }

      // Combine all chunk messages into a coherent commit message
      if (commitMessages.length === 0) {
        throw new Error('Failed to generate commit message from any chunk');
      }

      // For multiple chunks, create a summary message
      const combinedMessage = await this.combineChunkMessages(
        commitMessages,
        pattern,
        temperature,
        topP,
        presence,
        frequency,
        model,
        verbose
      );
      return combinedMessage;
    }
  }

  private async processSingleChunk(
    chunk: DiffChunk,
    pattern: string,
    temperature: number,
    topP: number,
    presence: number,
    frequency: number,
    model: string,
    verbose: boolean
  ): Promise<string> {
    if (verbose) {
      console.log(`Processing chunk with OpenAI API using pattern: ${pattern}`);
    }

    // Load the system prompt from the pattern file
    const systemPrompt = await this.loadPattern(pattern);
    if (!systemPrompt) {
      throw new Error(`Failed to load pattern: ${pattern}`);
    }

    if (verbose) {
      console.log(`Sending request to OpenAI API with model: ${model}`);
    }

    try {
      // Send request to OpenAI
      const response = await this.openai.chat.completions.create({
        model: model,
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: chunk.content },
        ],
        temperature: this.convertTemperature(temperature),
        top_p: this.convertTopP(topP),
        presence_penalty: this.convertPenalty(presence),
        frequency_penalty: this.convertPenalty(frequency),
      });

      if (verbose) {
        console.log('Received response from OpenAI API');
      }

      if (response.choices && response.choices[0] && response.choices[0].message) {
        return response.choices[0].message.content?.trim() || '';
      } else {
        throw new Error('No response received from OpenAI API');
      }
    } catch (error) {
      if (verbose) {
        console.log(`Error calling OpenAI API: ${error instanceof Error ? error.message : 'Unknown error'}`);
      }
      throw new Error(`Failed to generate commit message: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  private async combineChunkMessages(
    chunkMessages: string[],
    pattern: string,
    temperature: number,
    topP: number,
    presence: number,
    frequency: number,
    model: string,
    verbose: boolean
  ): Promise<string> {
    if (verbose) {
      console.log('Combining chunk messages into final commit message...');
    }

    // Load the system prompt from the pattern file
    const systemPrompt = await this.loadPattern(pattern);
    if (!systemPrompt) {
      throw new Error(`Failed to load pattern: ${pattern}`);
    }

    const combinedContent = chunkMessages.join('\n\n');
    const estimatedTokens = this.estimateTokenCount(systemPrompt) + this.estimateTokenCount(combinedContent);

    if (estimatedTokens > OpenAIService.MAX_CONTEXT_TOKENS && chunkMessages.length > 1) {
      if (verbose) {
        console.log('Context length exceeded, re-chunking summaries...');
      }

      const groupedSummaries: string[] = [];
      const currentGroup: string[] = [];
      let currentTokens = this.estimateTokenCount(systemPrompt);

      for (const msg of chunkMessages) {
        const msgTokens = this.estimateTokenCount(msg);
        if (currentTokens + msgTokens > OpenAIService.MAX_CONTEXT_TOKENS / 2 && currentGroup.length > 0) {
          const summary = await this.combineChunkMessages(
            currentGroup,
            pattern,
            temperature,
            topP,
            presence,
            frequency,
            model,
            verbose
          );
          groupedSummaries.push(summary);
          currentGroup.length = 0;
          currentTokens = this.estimateTokenCount(systemPrompt);
        }

        currentGroup.push(msg);
        currentTokens += msgTokens;
      }

      if (currentGroup.length > 0) {
        const summary = await this.combineChunkMessages(
          currentGroup,
          pattern,
          temperature,
          topP,
          presence,
          frequency,
          model,
          verbose
        );
        groupedSummaries.push(summary);
      }

      return await this.combineChunkMessages(
        groupedSummaries,
        pattern,
        temperature,
        topP,
        presence,
        frequency,
        model,
        verbose
      );
    }

    try {
      // Send request to OpenAI
      const response = await this.openai.chat.completions.create({
        model: model,
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: combinedContent },
        ],
        temperature: this.convertTemperature(temperature),
        top_p: this.convertTopP(topP),
        presence_penalty: this.convertPenalty(presence),
        frequency_penalty: this.convertPenalty(frequency),
      });

      if (response.choices && response.choices[0] && response.choices[0].message) {
        return response.choices[0].message.content?.trim() || '';
      } else {
        throw new Error('No response received from OpenAI API');
      }
    } catch (error) {
      // If combining fails, return the first chunk message as fallback
      if (verbose) {
        console.log(`Error calling OpenAI API: ${error instanceof Error ? error.message : 'Unknown error'}`);
        console.log('Failed to combine messages, using first chunk as fallback');
      }
      const fallback = chunkMessages[0].split(':').slice(1).join(':').trim();
      return fallback || chunkMessages[0];
    }
  }

  /**
   * Loads a pattern file from the patterns directory
   */
  private async loadPattern(patternName: string): Promise<string> {
    const patternPath = join(this.patternsDirectory, patternName, 'system.md');

    try {
      return await fs.readFile(patternPath, 'utf8');
    } catch {
      throw new Error(`Pattern file not found: ${patternPath}`);
    }
  }

  /**
   * Converts the temperature value from the CLI (0-2) to OpenAI API range (0-2)
   */
  private convertTemperature(temperature: number): number {
    // OpenAI uses 0-2 for temperature
    return Math.max(0, Math.min(temperature, 2));
  }

  /**
   * Converts the top-p value from the CLI (0-1) to OpenAI API range (0-1)
   */
  private convertTopP(topP: number): number {
    // OpenAI uses 0-1 for top-p
    return Math.max(0, Math.min(topP, 1));
  }

  /**
   * Converts the penalty value from the CLI (-2 to 2) to OpenAI API range (-2 to 2)
   */
  private convertPenalty(penalty: number): number {
    // OpenAI uses -2 to 2 for penalties
    return Math.max(-2, Math.min(penalty, 2));
  }

  /**
   * Estimates token count using a rough 4 chars per token heuristic
   */
  private estimateTokenCount(text: string): number {
    return Math.max(1, Math.floor(text.length / 4));
  }
}